### PR TITLE
Users unable to close positions on ME/SOL pool without harvesting first

### PR DESCRIPTION
**Pull request by pauldragonfly**
Description:
This PR fixes a constraint error that occurs when calling closePosition on a position that has uninitialized rewards. The error message indicates that the reward_owner_account.mint does not match the expected mint address stored in whirlpool.reward_infos[reward_index].mint. This issue arises because closePosition attempts to collect rewards for all possible reward indexes, including those that are not initialized.

**Problem Statement:**
When attempting to close a position using closePosition, a constraint error occurs if the whirlpool has uninitialized rewards. The error is due to the program trying to collect rewards for uninitialized reward indexes, resulting in mismatched mint addresses.

**Solution:**
Modify the getClosePositionIx method in whirlpool-impl.ts to skip any uninitialized rewards when constructing the instructions for collectReward. This is achieved by filtering out reward indexes where the mint is PublicKey.default, which indicates that the reward is uninitialized.

**Steps:**
Filtering Uninitialized Rewards: The code now filters out any rewards where info.mint equals PublicKey.default, which signifies that the reward has not been initialized. This ensures that only initialized rewards are processed.
Maintaining Correct Indexing: By mapping over rewardInfos with both info and index, we preserve the correct reward index when processing the rewards.

**Edge Case Handling:**
Uninitialized Rewards: Skipping uninitialized rewards prevents the constraint error caused by attempting to collect rewards that do not exist.
Initialized Rewards with Zero Emission: Rewards that are initialized but have zero emission are still processed. If there's any accumulated reward (e.g., from past emissions), it will be collected. This ensures that no rewards are missed.

**Why Edge Cases Are Not a Problem:**
No Loss of Rewards: Since we only skip uninitialized rewards, there is no risk of missing out on any accrued rewards. All initialized rewards, regardless of their emission status, are still collected.
Consistency with collectReward Behavior: The collectReward function already handles uninitialized rewards gracefully by not attempting to collect them. By aligning closePosition with this behavior, we ensure consistency across the SDK.
Successful Execution: With this change, calling closePosition no longer results in a constraint error, even when some rewards are uninitialized.

**Testing:**
Manual Testing: Verified that closePosition works as expected when:
All rewards are initialized.
Some rewards are uninitialized.
Rewards have zero emission.

### CODE REVIEW

File: legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts [hunk 1]
```diff
@@ -599,14 +598,18 @@ export class WhirlpoolImpl implements Whirlpool {
     const shouldDecreaseLiquidity = positionData.liquidity.gtn(0);
 
     const rewardsToCollect = this.data.rewardInfos
-      .filter((_, i) => {
+      .map((info, i) => ({ info, index: i }))
+      .filter(({ info, index }) => {
+        if (info.mint.equals(PublicKey.default)) {
```
Comment (odcheung) [inline legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts:603]: we can use PoolUtil.isRewardInitialized here

File: legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts [hunk 2]
```diff
@@ -599,14 +598,18 @@ export class WhirlpoolImpl implements Whirlpool {
     const shouldDecreaseLiquidity = positionData.liquidity.gtn(0);
 
     const rewardsToCollect = this.data.rewardInfos
-      .filter((_, i) => {
+      .map((info, i) => ({ info, index: i }))
+      .filter(({ info, index }) => {
```
Comment (odcheung) [inline legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts:602]: looks like we are mapping to get the index to provide filter with the index. Does the filter's index paramter help in this case?

**Review by odcheung — COMMENTED** [review]
nice! you found it

File: legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts (hunk 1 above)
Comment (pauldragonfly) [reply legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts]: Good point, fixed!

File: legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts (hunk 2 above)
Comment (pauldragonfly) [reply legacy-sdk/whirlpool/src/impl/whirlpool-impl.ts]: Also fixed

**Review by wjthieme — APPROVED** [review]
🚢🚢🚢

**Pull request merged**